teps/0089-nonfalsifiable-provenance-support.md

```diff
@@ -0,0 +1,251 @@
+
+### Installing SPIRE
+SPIRE runs as a Unix domain socket on the k8s node.
+We can use [spiffe-csi](https://github.com/spiffe/spiffe-csi) to mount the SPIRE socket into Pods as a `csi` type Volume so that we don't have to rely on the `hostPath` volume.
+Users will be responsible for installing this themselves.
+When creating Pods, we would automatically mount this volume in as appropriate.
+
+This volume mount would look something like this on an arbitrary Pod created by the controller:
+
+```yaml
+containers:
+- name: my-image
+    volumeMounts:
+    - name: spiffe-workload-api
+      mountPath: /spiffe-workload-api
+      readOnly: true
+    env:
+    - name: SPIFFE_ENDPOINT_SOCKET
+      value: unix:///spiffe-workload-api/spire-agent.sock
+  volumes:
+  - name: spiffe-workload-api
+    csi:
+      driver: "csi.spiffe.io"
+```
+
+
+## Proposed Solution
+
+### 1. Tekton Chains has no way to verify that the TaskRun it received wasn't modified by anybody other than Tekton, during or after execution
+
+The solution to this is Signed TaskRuns, where the TaskRun has to be signed and verified every time it has been modified.
+That way, we can prove that the TaskRun hasn't been tampered with during or after execution.
+
+The Tekton Controller will need to request a signature from SPIRE each time the TaskRun is modified to prove that it hasn't been tampered with during execution.
+Roughly, this will look something like this:
+* Controller initiates a TaskRun, and stores a signature over its contents as an annotation
+
+Each time the TaskRun is modified,
+* Controller verifies the TaskRun hasn't been modified
+* Controller requests a new SPIRE SVID and signature over the new modified TaskRun
+* Controller stores the new SVID and signature on the TaskRun as an annotation
+
+Right now, the design details around this are a little fuzzy!
+The plan is to first implement signed Results, and then if this design no longer seems like the correct fit for signed TaskRuns then it will be revisited in this TEP :)
+
+Things to keep in mind when designing this feature further:
+* There could potentially be a performance implication here, if we need to request signatures every time the TaskRun is modified during execution.
+* We may want to ignore labels/annotations when checking for modifications, since they might be updated by other services
+* What if mutating admission controllers are intentionally changing TaskRuns (e.g. Solarwinds injecting Tasks into a Pipeline)
+
+### 2. Tekton Pipelines can't verify that the results it reads weren't modified
+The solution to this is Signed Results.
+We will modifiy the entrypointer image to sign results with SPIRE once they're available.
```

Comment:
Do you know how fine grained authorization is for the signing keys? e.g. does it allow per-container authorization or is it only for the Pod? I'm wondering if/how we might be able to reduce access to keys from user steps 🤔

Reply:
There are a few different [selectors](https://github.com/spiffe/spire/blob/main/doc/plugin_agent_workloadattestor_k8s.md) you can configure, including the container image and container name, so I believe we should be able to reduce access from user steps.